src/storage/v2/inmemory/edge_type_index.cpp

```diff
@@ -17,66 +17,200 @@
-                                        std::optional<SnapshotObserverInfo> const &snapshot_info) {
-  auto [it, emplaced] = index_.try_emplace(edge_type);
-  if (!emplaced) {
-    return false;
+namespace {
+inline void TryInsertEdgeTypeIndex(Vertex &from_vertex, EdgeTypeId edge_type, auto &&index_accessor,
+                                   std::optional<SnapshotObserverInfo> const &snapshot_info) {
+  if (from_vertex.deleted) {
+    return;
   }
 
-  utils::MemoryTracker::OutOfMemoryExceptionEnabler oom_exception;
-  try {
-    auto edge_acc = it->second.access();
-    for (auto &from_vertex : vertices) {
-      if (from_vertex.deleted) {
-        continue;
-      }
+  for (auto const &[type, to_vertex, edge_ref] : from_vertex.out_edges) {
+    if (type != edge_type) continue;
+    if (to_vertex->deleted) {
+      continue;
+    }
+    index_accessor.insert({&from_vertex, to_vertex, edge_ref.ptr, 0});
+    if (snapshot_info) {
+      snapshot_info->Update(UpdateType::EDGES);
+    }
+  }
+}
 
-      for (auto &edge : from_vertex.out_edges) {
-        const auto type = std::get<kEdgeTypeIdPos>(edge);
-        if (type == edge_type) {
-          auto *to_vertex = std::get<kVertexPos>(edge);
-          if (to_vertex->deleted) {
-            continue;
-          }
-          edge_acc.insert({&from_vertex, to_vertex, std::get<kEdgeRefPos>(edge).ptr, 0});
-          if (snapshot_info) {
-            snapshot_info->Update(UpdateType::EDGES);
-          }
-        }
-      }
+inline void TryInsertEdgeTypeIndex(Vertex &from_vertex, EdgeTypeId edge_type, auto &&index_accessor,
+                                   std::optional<SnapshotObserverInfo> const &snapshot_info, Transaction const &tx) {
+  bool exists = true;
+  bool deleted = false;
+  Delta *delta = nullptr;
+  utils::small_vector<Vertex::EdgeTriple> edges;
+  auto matches_edge_type = [edge_type](auto const &each) { return std::get<EdgeTypeId>(each) == edge_type; };
+  {
+    auto guard = std::shared_lock{from_vertex.lock};
+    deleted = from_vertex.deleted;
+    delta = from_vertex.delta;
+    edges = from_vertex.out_edges | rv::filter(matches_edge_type) | r::to<utils::small_vector<Vertex::EdgeTriple>>;
+  }
+  // Create and drop index will always use snapshot isolation
+  if (delta) {
+    ApplyDeltasForRead(&tx, delta, View::OLD, [&](const Delta &delta) {
+      // clang-format off
+      DeltaDispatch(delta, utils::ChainedOverloaded{
+        Exists_ActionMethod(exists),
+        Deleted_ActionMethod(deleted),
+        Edges_ActionMethod<EdgeDirection::OUT>(edges, edge_type)
+      });
+      // clang-format on
+    });
+  }
+  if (!exists || deleted || edges.empty()) {
+    return;
+  }
+
+  for (auto const &[type, to_vertex, edge_ref] : edges) {
+    index_accessor.insert({&from_vertex, to_vertex, edge_ref.ptr, tx.start_timestamp});
+    if (snapshot_info) {
+      snapshot_info->Update(UpdateType::EDGES);
+    }
+  }
+}
```

Comment:
@as51340 this is new code by me `snapshot_info->Update(UpdateType::EDGES)` should it be done once per edge?